packages/plugin-ext/src/main/browser/editor/untitled-resource.ts

```diff
@@ -14,39 +14,89 @@
-        if (uri.scheme === Schemes.UNTITLED) {
-            return resources.get(uri.toString())!;
+
+    @inject(FileResourceResolver)
+    protected readonly fileResourceResolver: FileResourceResolver;
+
+    protected readonly resources = new Map<string, UntitledResource>();
+
+    async resolve(uri: URI): Promise<UntitledResource> {
+        if (uri.scheme !== Schemes.UNTITLED) {
+            throw new Error('The given uri is not untitled file uri: ' + uri);
+        } else {
+            const untitledResource = this.resources.get(uri.toString());
+            if (!untitledResource) {
+                return this.createUntitledResource('', '', uri, this.fileResourceResolver);
+            } else {
+                return untitledResource;
+            }
+        }
+    }
+
+    async  createUntitledResource(content?: string, language?: string, uri?: URI, fileResourceResolver?: FileResourceResolver): Promise<UntitledResource> {
+        let extension;
+        if (language) {
+            for (const lang of monaco.languages.getLanguages()) {
+                if (lang.id === language) {
+                    if (lang.extensions) {
+                        extension = lang.extensions[0];
+                        break;
+                    }
+                }
+            }
         }
-        throw new Error(`scheme ${uri.scheme} is not '${Schemes.UNTITLED}'`);
+        return new UntitledResource(this.resources, uri ? uri : new URI().withScheme(Schemes.UNTITLED).withPath(`/Untitled-${index++}${extension ? extension : ''}`),
+            content, fileResourceResolver);
     }
 }
 
 export class UntitledResource implements Resource {
+    private fileResource?: FileResource;
 
-    constructor(public uri: URI, private content?: string) {
-        resources.set(this.uri.toString(), this);
+    constructor(private resources: Map<string, UntitledResource>, public uri: URI, private content?: string, private fileResourceResolver?: FileResourceResolver) {
+        this.resources.set(this.uri.toString(), this);
     }
 
-    readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
-        return Promise.resolve(this.content ? this.content : '');
+    dispose(): void {
+        this.resources.delete(this.uri.toString());
     }
 
-    dispose(): void {
-        resources.delete(this.uri.toString());
+    async readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
+        if (this.fileResource) {
+            return this.fileResource.readContents(options);
+        } else if (this.content) {
+            return Promise.resolve(this.content);
+        } else {
+            return '';
+        }
+    }
+
+    async saveContents(content: string, options?: { encoding?: string, overwriteEncoding?: string }): Promise<void> {
+        if (this.fileResourceResolver) {
+            this.fileResource = await this.fileResourceResolver.resolve(new URI(this.uri.path.toString()));
```

Comment:
we shoudl not create file resource on each save, only on first

Reply:
done